src/main/protobuf/bazel_output_service.proto

```diff
@@ -0,0 +1,324 @@
+    google.rpc.Status status = 1;
+  }
+
+  // The response for each of the requested artifacts, using the same order as
+  // requested. This means that this list has the same length as
+  // StageArtifactsRequest.artifacts.
+  repeated Response responses = 1;
+}
+
+message FinalizeArtifactsRequest {
+  message Artifact {
+    // path is relative to StartBuildResponse.output_path.
+    string path = 1;
+    // Expected digest for this file. This allows server to detect if the file
+    // has been changed after Bazel finished creating the file and the
+    // corresponding FinalizeArtifactsRequest is processed. This field is
+    // optional for regular files, and should be omitted for directories and
+    // symlinks.
+    //
+    // The concrete type of the digest depending on the CAS Bazel connects to.
+    // In case of a REv2 CAS, the type is
+    // [Digest][build.bazel.remote.execution.v2.Digest].
+    google.protobuf.Any digest = 2;
+  }
+
+  // The identifier of the build. Server uses this to determine which output
+  // path needs to be modified.
+  string build_id = 1;
+
+  repeated Artifact artifacts = 2;
+}
+
+message FinalizeArtifactsResponse {
+  // Intentionally empty for future extensibility.
+}
+
+message FinalizeBuildRequest {
+  // The identifier of the build that should be finalized.
+  string build_id = 1;
+
+  // Whether the build completed successfully. The remote output service may,
+  // for example, use this option to apply different retention policies that
+  // take the outcome of the build into account.
+  bool build_successful = 2;
+}
+
+message FinalizeBuildResponse {
+  // Intentionally empty for future extensibility.
+}
+
+message BatchStatRequest {
+  // The identifier of the build. The remote output service uses this to
+  // determine which output path needs to be inspected.
+  string build_id = 1;
+
+  // Paths whose status needs to be obtained. The server MUST resolve the path
+  // using lstat semantic, i.e. components except the last one must be resolved
+  // if they are symlinks.
+  //
+  // Path is relative to StartBuildResponse.output_path.
+  repeated string paths = 2;
+}
+
+message BatchStatResponse {
+  message StatResponse {
+    // The status of the file. If the file corresponding with the requested path
+    // does not exist, this field will be null.
+    FileStat file_stat = 1;
+  }
+
+  message FileStat {
+    message File {
+      // The digest of the file.
+      //
+      // The concrete type of the digest depending on the CAS Bazel connects to.
+      // In case of a REv2 CAS, the type is
+      // [Digest][build.bazel.remote.execution.v2.Digest].
+      google.protobuf.Any digest = 1;
+    }
+
+    message Symlink {
+      // The target of the symbolic link.
+      string target = 1;
+    }
+
+    message Directory {
+    }
+
+    // If the path resolves to a location outside the output path where the
+    // server is unable to obtain its status, none of the file_type will be set.
```

Comment:
In addition to the canonical path pointing outside of the output path, there's also the case where it points inside the output path, but at least one of the symlinks whose resolution was required pointed outside of it.

For example, consider path `bazel-out/cfg/bin/a/b`, where `.../a` is a symlink to `/foo` (outside of the output path), and `/foo` is a symlink to `.../bazel-out/cfg/bin/c`. The path does canonicalize to a location inside the output path, but we had to peek outside of it to arrive at the answer.

It's also unclear whether we're saying "the server MUST NOT attempt to resolve any such path" or "the server MAY attempt to resolve such a path, but is not required to" (because it might not have access, etc). In the presence of filesystem namespaces, only the former is safe (i.e., in the example above, `/foo` might point to different locations as seen by the server and client). This leads to wording such as the following:

If, in the process of canonicalizing the path according to lstat semantics, a symlink pointing to a location outside of the output path must be followed, the server is not required to provide a response, and MUST NOT set any of the `file_type` fields.

(Or, if we do want to keep `output_path_aliases`, we have to modify the above to precisely state how it affects the outcome of path canonicalization.)